staging/src/k8s.io/dynamic-resource-allocation/controller/controller.go

```diff
@@ -0,0 +1,777 @@
+					return fmt.Errorf("allocation of pod claim %s failed: %v", delayed.PodClaimName, err)
+				}
+			}
+		}
+	}
+
+	// Now update unsuitable nodes. This is useful information for the scheduler even if
+	// we managed to allocate because we might have to undo that.
+	// TODO: replace with patching the array. We can do that without race conditions
+	// because each driver is responsible for its own entries.
+	modified := false
+	podScheduling = podScheduling.DeepCopy()
+	for _, delayed := range claims {
+		i := findClaim(podScheduling.Status.Claims, delayed.PodClaimName)
+		if i < 0 {
+			// Add new entry.
+			podScheduling.Status.Claims = append(podScheduling.Status.Claims,
+				v1.ResourceClaimSchedulingStatus{
+					PodResourceClaimName: delayed.PodClaimName,
+					UnsuitableNodes:      delayed.UnsuitableNodes,
+				})
+			modified = true
+		} else if stringsDiffer(podScheduling.Status.Claims[i].UnsuitableNodes, delayed.UnsuitableNodes) {
+			// Update existing entry.
+			podScheduling.Status.Claims[i].UnsuitableNodes = delayed.UnsuitableNodes
+			modified = true
+		}
+	}
+	if modified {
+		logger.V(6).Info("Updating pod scheduling with modified unsuitable nodes", "podScheduling", podScheduling)
+		if _, err := ctrl.kubeClient.CoreV1().PodSchedulings(podScheduling.Namespace).UpdateStatus(ctx, podScheduling, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("update unsuitable node status: %v", err)
+		}
+	}
+
+	// We must keep the object in our queue and keep updating the
+	// UnsuitableNodes fields.
+	return errPeriodic
+}
+
+type claimAllocations []*ClaimAllocation
+
+// MarshalLog replaces the pointers with the actual structs because
+// we care about the content, not the pointer values.
+func (claims claimAllocations) MarshalLog() interface{} {
+	content := make([]ClaimAllocation, 0, len(claims))
+	for _, claim := range claims {
+		content = append(content, *claim)
+	}
+	return content
+}
+
+var _ logr.Marshaler = claimAllocations{}
+
+func findClaim(claims []v1.ResourceClaimSchedulingStatus, podClaimName string) int {
+	for i := range claims {
+		if claims[i].PodResourceClaimName == podClaimName {
+			return i
+		}
+	}
+	return -1
+}
+
+func hasString(strings []string, str string) bool {
+	for _, s := range strings {
+		if s == str {
+			return true
+		}
+	}
+	return false
+}
+
+// stringsDiffer does a strict comparison of two string arrays, order of entries matters.
+func stringsDiffer(a, b []string) bool {
+	if len(a) != len(b) {
+		return true
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return true
+		}
+	}
+	return false
+}
+
+func (ctrl *controller) hasFinalizer(claim *v1.ResourceClaim) bool {
+	for _, finalizer := range claim.Finalizers {
+		if finalizer == ctrl.finalizer {
+			return true
+		}
+	}
+	return false
+}
+
+func (ctrl *controller) removeFinalizer(in []string) []string {
```

Comment:
nit: add doc string here for consistency with all other helpers above and below

Reply:
Done.